Support for julia 1.12

VarLad opened the pull request:
Based on https://github.com/tshort/StaticCompiler.jl/pull/174
~~Currently filing this as a draft because Bumper.jl integration seems to be broken.~~

kylincaster committed:
update to julia1.11 and GPUCompiler 1.5.2

Commit:
update Project.toml

Commit:
pass the test for julia 1.8~1.11 in ubuntu

Commit:
update the version to 0.7.3

Commit:
update for GPUCompiler by removing the deprecated API

VarLad committed:
initial 1.12 support

MasonProtter commented on src/interpreter.jl:
```diff
@@ -1,61 +1,68 @@
+        token::Any
+    else
+        code_cache::CodeCache # global cache
+    end
```
moving the compat bounds up to v1.11 would help us avoid this logic

MasonProtter commented on src/quirks.jl:
```diff
@@ -1,10 +1,41 @@
     quote
-        println($err)
+        printf($err)
         libcexit(Int32(1))
```
Why was this changed?

VarLad replied:
from previous pr
reverted

MasonProtter commented on test/runtests.jl:
```diff
@@ -6,18 +6,18 @@ using LoopVectorization
 
 addprocs(1)
-@everywhere using StaticCompiler, StrideArraysCore
+# @everywhere using StaticCompiler, StrideArraysCore
```
Why is this removed?

VarLad replied:
from previous PR
reverted

MasonProtter commented on test/scripts/loopvec_matrix.jl:
```diff
@@ -3,9 +3,11 @@ using StaticTools
-        for k ∈ indices((A,B), (2,1))
+        # for k ∈ indices((A,B), (2,1))
+        for k ∈ indices(A, 2)
             Cmn += A[m,k] * B[k,n]
```
Why is this removed? Was it to avoid boundserrors causing it to fail to compile? If so, we should see if there's a device_override we could do to help the compiler not get tripped up here.

MasonProtter commented on test/scripts/loopvec_matrix.jl:
```diff
@@ -39,7 +41,8 @@ function loopvec_matrix(argc::Int, argv::Ptr{Ptr{UInt8}})
 
     # Print to stdout
-    printf(C)
+    printf(c"C matric = \n")
```
```suggestion
    printf(c"C matrix = \n")
```

MasonProtter reviewed (changes requested):
Thanks for trying to revive this @VarLad. I mostly have the same comment as I did for the previous one, so I'll just copy-paste it:

> I think it probably doesn't make sense to try and have so many `if version(...)` checks in here, and we should just increase our version support bounds. 
>
> If people aren't able to upgrade to newer versions of julia or GPUCompiler, they will get stuck with older versions of StaticCompiler which is fine.
>
> Also, the `pkgversion` function was added in v1.9, so using it will break the v1.8 tests. 
>
> Since v1.10 is the current julia LTS, I say we drop support for v1.8 and 1.9, and drop support for GPUCompiler < 1.5
>
> That should make this simpler and hopefully more robust too. 

Except I'd recommend now that we also drop v1.10 support and move onto v1.11/v1.12 support since there's already a released version that works on v1.10 anyways, and this isn't adding new features that people on 1.10 would want (if anything, it seems that some features have regressed, since it seems you changed some of the tests to remove the possibility of them emitting boundserrors)

MasonProtter commented on src/quirks.jl:
```diff
@@ -37,9 +68,12 @@ end
+if VERSION < v"1.11"
+    @device_override @noinline Base.throw_boundserror(A, I) =
+        @print_and_throw c"Out-of-bounds array access"
+end
```
Oh, I see you also kept this from the old PR. Copying my comment from the previous time I reviewed this:

> `Base.throw_boundserror` definitely was not removed in v1.11. 

Please restore the `throw_boundserror` overrides, that's likely what's causing the failures you saw related to matrices and why you had to remove the boundschecking from the tests.

MasonProtter commented:
In the future, if you're reviving old PRs, please make sure you address the review feedback that's already been given on the older version of the PR before re-requesting that someone review it. Because of the amount of time that had past, I didn't immediately recognize that I had already made the exact same review comments back in June, so I spent some time basically just rediscovering the same complaints I already had back then.

VarLad commented on test/testcore.jl:
```diff
@@ -10,7 +8,7 @@ fib(n) = n <= 1 ? n : fib(n - 1) + fib(n - 2) # This needs to be defined globall
 
     #Compile dylib
-    name = repr(fib)
+    name = string(nameof(fib)) # repr(fib)
```
This is what I see if I use `repr`:
```
     Testing Running tests...
Standalone Dylibs: Error During Test at ~/.julia/dev/StaticCompiler/test/testcore.jl:5
  Got exception outside of a @test
  UndefVarError: `name` not defined in `Main`
  Suggestion: check for spelling errors or missing imports.
  Hint: a global variable of this name also exists in LLVM.
```

MasonProtter replied:
Strange

VarLad committed:
address reviews

VarLad committed:
re-enable Bumper tests, comment printdlm

VarLad commented:
@MasonProtter apologies for wasting time, I forgot about the reviews in last PR. I'll be careful about it in future PRs. :smiling_face_with_tear: 

I've addressed most of the comments. I've removed explicit support for anything pre-1.11 from the codebase. 
I've reverted the loopvec based tests (they were failing before, and are still failing) 

printdlm function call fails test as well with:
```
/usr/bin/ld: ./times_table.o: in function `times_table':
start:(.text+0x330): undefined reference to `ijl_error'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

VarLad committed:
revert printf

brenhinkeller commented:
> There are a lot of warnings about:
> 
> ```
> WARNING: llvmcall with integer pointers is deprecated.
> Use actual pointers instead, replacing i32 or i64 with i8* or ptr
> ```
> 
> throughout the codebase 🤔

Yeah, this is probably due to the switch to opaque pointers in LLVM, and will require a bunch of changes in StaticTools, but I think they should be relatively straightforward/repetitive

VarLad committed:
use 1.11 and 1.12 in ci

VarLad committed:
mark loopvec tests as broken

MasonProtter commented on src/interpreter.jl:
```diff
@@ -1,61 +1,53 @@
     opt_params::OptimizationParams
-
-    function StaticInterpreter(cache::CodeCache, mt::Union{Nothing,Core.MethodTable}, world::UInt, ip::InferenceParams, op::OptimizationParams)
+    # token_or_cache = token::Any, code_cache::CodeCache
```
Is this comment here for a reason?

MasonProtter reviewed (commented):
Almost there! Thanks for sticking with it, just a few nitpicks then we can merge it.

VarLad committed:
address comments

VarLad commented on src/interpreter.jl:
```diff
@@ -1,61 +1,53 @@
+
+const CodeCache = Nothing
+# https://github.com/JuliaGPU/GPUCompiler.jl/src/jlgen.jl8#L322
+# as from struct GPUInterpreter <: CC.AbstractInterpreter 
```
Removed

VarLad commented on test/testintegration.jl:
```diff
@@ -205,13 +203,13 @@ end
         A = (1:10) * (1:5)'
         # Check ascii output
-        # @test parsedlm(c"table.tsv",'\t') == A' * A broken=Sys.isapple()
+        # @test_broken parsedlm(c"table.tsv",'\t') == A' * A broken=Sys.isapple()
```
From previous PR, reverted

VarLad commented:
This is quite weird, StaticCompiler both builds and all tests (pass + broken) work for me locally :thinking:

VarLad committed:
comment printdlm again
